File: js/analytics.js
(function () {
  'use strict';

  if (window.__vnfestAnalyticsTracked) return;
  window.__vnfestAnalyticsTracked = true;

  // This collector is intentionally independent from the site's language,
  // theme, and header runtimes. A failed analytics request must never affect
  // the page that the visitor is trying to use.
  var script = document.currentScript;
  var endpoint = '';
  try {
    if (!script) {
      var scripts = document.getElementsByTagName('script');
      for (var i = scripts.length - 1; i >= 0; i -= 1) {
        if (/\/js\/analytics\.js(?:\?|$)/.test(scripts[i].src || '')) {
          script = scripts[i];
          break;
        }
      }
    }
    endpoint = new URL('../api/analytics.php?action=track', (script && script.src) || window.location.href).href;
  } catch (error) {
    endpoint = '/api/analytics.php?action=track';
  }

  function randomUuid() {
    try {
      if (window.crypto && typeof window.crypto.randomUUID === 'function') return window.crypto.randomUUID();
      if (window.crypto && typeof window.crypto.getRandomValues === 'function') {
        var bytes = new Uint8Array(16);
        window.crypto.getRandomValues(bytes);
        bytes[6] = (bytes[6] & 15) | 64;
        bytes[8] = (bytes[8] & 63) | 128;
        var hex = Array.prototype.map.call(bytes, function (byte) {
          return ('0' + byte.toString(16)).slice(-2);
        }).join('');
        return hex.slice(0, 8) + '-' + hex.slice(8, 12) + '-4' + hex.slice(13, 16) + '-' + ((parseInt(hex.slice(16, 18), 16) & 3 | 8).toString(16)) + hex.slice(17, 20) + '-' + hex.slice(20);
      }
    } catch (error) {}
    return '00000000-0000-4000-8000-' + String(Date.now()).slice(-12).padStart(12, '0');
  }

  function getCookie(name) {
    var prefix = name + '=';
    var cookies = document.cookie ? document.cookie.split(';') : [];
    for (var i = 0; i < cookies.length; i += 1) {
      var value = cookies[i].replace(/^\s+/, '');
      if (value.indexOf(prefix) === 0) return decodeURIComponent(value.slice(prefix.length));
    }
    return '';
  }

  function setCookie(name, value) {
    var cookie = name + '=' + encodeURIComponent(value) + '; Max-Age=31536000; Path=/; SameSite=Lax';
    if (window.location.protocol === 'https:') cookie += '; Secure';
    document.cookie = cookie;
  }

  function visitorId() {
    var current = getCookie('vnfest_visitor_id');
    if (/^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(current)) return current;
    current = randomUuid();
    setCookie('vnfest_visitor_id', current);
    return current;
  }

  function hostOf(referrer) {
    try { return referrer ? new URL(referrer).hostname.toLowerCase() : ''; } catch (error) { return ''; }
  }

  function sourceCategory(referrerHost) {
    if (!referrerHost) return 'direct';
    if (referrerHost === window.location.hostname.toLowerCase()) return 'internal';
    if (/(^|\.)(google|bing|baidu|sogou|yahoo|duckduckgo)\./i.test(referrerHost)) return 'search';
    if (/(^|\.)(twitter|x|facebook|instagram|youtube|bilibili|weibo|reddit|discord)\./i.test(referrerHost)) return 'social';
    return 'external';
  }

  function deviceType() {
    var ua = navigator.userAgent || '';
    if (/iPad|Tablet|Android(?!.*Mobile)/i.test(ua)) return 'tablet';
    if (/Android|iPhone|iPod|Mobile|Windows Phone/i.test(ua)) return 'mobile';
    if (ua) return 'desktop';
    return 'unknown';
  }

  function browserName() {
    var ua = navigator.userAgent || '';
    if (/Edg\//i.test(ua)) return 'edge';
    if (/Firefox\//i.test(ua)) return 'firefox';
    if (/Chrome\//i.test(ua) || /Chromium\//i.test(ua)) return 'chrome';
    if (/Safari\//i.test(ua)) return 'safari';
    return 'other';
  }

  function shouldSkip(path) {
    return !/^https?:$/i.test(window.location.protocol)
      || /\/(?:admin|api|scripts|includes|data|uploads|node_modules|vendor)(?:\/|$)/i.test(path)
      || /(?:^|\/)(?:test|tests|fixture|fixtures)(?:\/|[-_.]|$)/i.test(path);
  }

  function send(payload) {
    var serialized = JSON.stringify(payload);
    try {
      if (navigator.sendBeacon) {
        var blob = new Blob([serialized], { type: 'application/json' });
        if (navigator.sendBeacon(endpoint, blob)) return;
      }
    } catch (error) {}
    try {
      fetch(endpoint, {
        method: 'POST',
        credentials: 'same-origin',
        headers: { 'Content-Type': 'application/json' },
        body: serialized,
        keepalive: true
      }).catch(function () {});
    } catch (error) {}
  }

  try {
    var path = window.location.pathname || '/';
    if (shouldSkip(path)) return;
    var referrerHost = hostOf(document.referrer || '');
    send({
      event_id: randomUuid(),
      visitor_id: visitorId(),
      page_path: path,
      page_title: document.title || '',
      source_category: sourceCategory(referrerHost),
      referrer_host: referrerHost,
      device_type: deviceType(),
      browser_name: browserName()
    });
  } catch (error) {}
})();

File: backend/internal/jobs/analytics.go
package jobs

import (
	"bufio"
	"compress/gzip"
	"context"
	"fmt"
	"io"
	"net"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/VNFestMap/galgame-community-map/backend/internal/config"
	"github.com/VNFestMap/galgame-community-map/backend/internal/store/sqlstore"
)

type AnalyticsBackfillOptions struct {
	From, To string
	Until    *time.Time
	Logs     []string
	DryRun   bool
}

type AnalyticsBackfillReport struct {
	Success             bool     `json:"success"`
	DryRun              bool     `json:"dry_run"`
	From                string   `json:"from"`
	To                  string   `json:"to"`
	LogFiles            []string `json:"log_files"`
	LinesRead           int      `json:"lines_read"`
	MatchedPageRequests int      `json:"matched_page_requests"`
	AggregateRows       int      `json:"aggregate_rows"`
	SkippedLines        int      `json:"skipped_lines"`
	HistoricalUV        *int     `json:"historical_uv"`
}

var historicalAccessLogRE = regexp.MustCompile(`^\S+\s+\S+\s+\S+\s+\[(?P<time>[^\]]+)\]\s+"(?P<request>[^"]*)"\s+(?P<status>\d{3})\s+\S+\s+"(?P<referrer>[^"]*)"\s+"(?P<ua>[^"]*)"`)

type historicalPVKey struct {
	day, page, title, source, host, device, browser string
}

// RunAnalyticsBackfill ports the privacy-preserving historical PV importer.
// It never manufactures UV and never stores IP addresses, raw user agents, or
// complete referrer URLs.
func RunAnalyticsBackfill(ctx context.Context, db *sqlstore.DB, cfg config.Config, options AnalyticsBackfillOptions) (AnalyticsBackfillReport, error) {
	from := options.From
	if from == "" {
		from = "2026-08-10"
	}
	to := options.To
	if to == "" {
		to = "2026-09-05"
	}
	if !validDateKey(from) || !validDateKey(to) || from > to {
		return AnalyticsBackfillReport{}, fmt.Errorf("invalid from/to; use YYYY-MM-DD")
	}
	exists, err := db.TableExists(ctx, "analytics_historical_pv")
	if err != nil {
		return AnalyticsBackfillReport{}, err
	}
	if !exists {
		return AnalyticsBackfillReport{}, fmt.Errorf("%w: analytics_historical_pv", ErrSchemaUnavailable)
	}
	logs := options.Logs
	if len(logs) == 0 {
		logs = []string{"/www/wwwlogs/162.251.93.178.log.20260903-132957.gz", "/www/wwwlogs/162.251.93.178.log"}
	}
	report := AnalyticsBackfillReport{Success: true, DryRun: options.DryRun, From: from, To: to, LogFiles: append([]string(nil), logs...), HistoricalUV: nil}
	aggregate := map[historicalPVKey]int{}
	for _, logPath := range logs {
		file, closeFn, openErr := openHistoricalLog(logPath)
		if os.IsNotExist(openErr) {
			continue
		}
		if openErr != nil {
			return report, fmt.Errorf("open historical log: %w", openErr)
		}
		scanner := bufio.NewScanner(file)
		buffer := make([]byte, 64*1024)
		scanner.Buffer(buffer, 2<<20)
		for scanner.Scan() {
			report.LinesRead++
			line := scanner.Text()
			match := historicalAccessLogRE.FindStringSubmatch(line)
			if match == nil {
				report.SkippedLines++
				continue
			}
			fields := map[string]string{}
			for index, name := range historicalAccessLogRE.SubexpNames() {
				if index > 0 && name != "" && index < len(match) {
					fields[name] = match[index]
				}
			}
			requestParts := strings.Fields(fields["request"])
			if len(requestParts) < 2 || strings.ToUpper(requestParts[0]) != "GET" {
				report.SkippedLines++
				continue
			}
			requestTime, parseErr := parseAccessTime(fields["time"])
			if parseErr != nil {
				report.SkippedLines++
				continue
			}
			if options.Until != nil && !requestTime.Before(*options.Until) {
				report.SkippedLines++
				continue
			}
			day := requestTime.In(mustLocationJob("Asia/Shanghai")).Format("2006-01-02")
			if day < from || day > to {
				report.SkippedLines++
				continue
			}
			status := 0
			_, _ = fmt.Sscanf(fields["status"], "%d", &status)
			if status < 200 || status >= 400 {
				report.SkippedLines++
				continue
			}
			page := normalizeHistoricalPath(requestParts[1])
			if page == "" || (page != "/" && !strings.HasSuffix(strings.ToLower(page), ".html")) {
				report.SkippedLines++
				continue
			}
			source, host := historicalSourceCategory(fields["referrer"], cfg.SiteURL)
			key := historicalPVKey{day: day, page: page, title: historicalPageTitle(cfg.Root, page), source: source, host: host, device: historicalDevice(fields["ua"]), browser: historicalBrowser(fields["ua"])}
			aggregate[key]++
			report.MatchedPageRequests++
		}
		if err := scanner.Err(); err != nil {
			_ = closeFn()
			return report, fmt.Errorf("read historical log: %w", err)
		}
		if err := closeFn(); err != nil {
			return report, err
		}
	}
	report.AggregateRows = len(aggregate)
	if options.DryRun || len(aggregate) == 0 {
		return report, nil
	}
	return report, writeHistoricalPV(ctx, db, aggregate)
}

func validDateKey(value string) bool {
	_, err := time.Parse("2006-01-02", value)
	return err == nil
}

func parseAccessTime(value string) (time.Time, error) {
	return time.Parse("02/Jan/2006:15:04:05 -0700", value)
}

func openHistoricalLog(path string) (io.ReadCloser, func() error, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, func() error { return nil }, err
	}
	if strings.HasSuffix(strings.ToLower(path), ".gz") {
		reader, gzipErr := gzip.NewReader(file)
		if gzipErr != nil {
			_ = file.Close()
			return nil, func() error { return nil }, gzipErr
		}
		return reader, func() error {
			closeErr := reader.Close()
			fileErr := file.Close()
			if closeErr != nil {
				return closeErr
			}
			return fileErr
		}, nil
	}
	return file, file.Close, nil
}

func normalizeHistoricalPath(value string) string {
	if parsed, err := url.Parse(value); err == nil && parsed.IsAbs() {
		value = parsed.Path
	}
	if index := strings.IndexAny(value, "?#"); index >= 0 {
		value = value[:index]
	}
	value = strings.ReplaceAll(strings.TrimSpace(value), "\\", "/")
	if value == "" {
		value = "/"
	}
	if !strings.HasPrefix(value, "/") {
		value = "/" + value
	}
	for strings.Contains(value, "//") {
		value = strings.ReplaceAll(value, "//", "/")
	}
	if historicalForbiddenPathRE.MatchString(value) || historicalFixturePathRE.MatchString(value) {
		return ""
	}
	return value
}

var historicalForbiddenPathRE = regexp.MustCompile(`(?i)/(?:admin|api|scripts|includes|data|uploads|node_modules|vendor)(?:/|$)`)
var historicalFixturePathRE = regexp.MustCompile(`(?i)(?:^|/)(?:test|tests|fixture|fixtures)(?:/|[-_.]|$)`)

func historicalPageTitle(root, page string) string {
	relative := page
	if page == "/" {
		relative = "/index.html"
	}
	file := filepath.Join(root, filepath.FromSlash(strings.TrimPrefix(relative, "/")))
	data, err := os.ReadFile(file)
	if err != nil {
		return ""
	}
	titleRE := regexp.MustCompile(`(?is)<title[^>]*>(.*?)</title>`)
	match := titleRE.FindSubmatch(data)
	if len(match) < 2 {
		return ""
	}
	title := regexp.MustCompile(`(?s)<[^>]+>`).ReplaceAllString(string(match[1]), "")
	return trimHistoricalText(title, 255)
}

func trimHistoricalText(value string, max int) string {
	value = strings.Map(func(r rune) rune {
		if r < 0x20 || r == 0x7f {
			return -1
		}
		return r
	}, strings.TrimSpace(value))
	runes := []rune(value)
	if len(runes) > max {
		return string(runes[:max])
	}
	return value
}

func historicalSourceCategory(referrer, siteURL string) (string, string) {
	referrer = strings.TrimSpace(referrer)
	if referrer == "" || referrer == "-" {
		return "direct", ""
	}
	parsed, err := url.Parse(referrer)
	if err != nil {
		return "external", ""
	}
	host := normalizeHistoricalHost(parsed.Hostname())
	if host == "" {
		return "external", ""
	}
	siteHost := ""
	if site, siteErr := url.Parse(siteURL); siteErr == nil {
		siteHost = normalizeHistoricalHost(site.Hostname())
	}
	if host == siteHost || host == "map.vnfest.top" || host == "www.map.vnfest.top" {
		return "internal", host
	}
	if strings.Contains(host, "google.") || strings.Contains(host, "bing.") || strings.Contains(host, "baidu.") || strings.Contains(host, "sogou.") || strings.Contains(host, "duckduckgo.") || strings.Contains(host, "yahoo.") || strings.Contains(host, "yandex.") {
		return "search", host
	}
	for _, social := range []string{"facebook.", "instagram.", "twitter.", "weibo.", "weixin.", "wechat.", "qq.com", "douban.", "bilibili.", "discord.", "reddit.", "x.com"} {
		if strings.Contains(host, social) {
			return "social", host
		}
	}
	return "external", host
}

func normalizeHistoricalHost(value string) string {
	value = strings.ToLower(strings.Trim(strings.TrimSpace(value), "."))
	if value == "" || len(value) > 253 || net.ParseIP(value) != nil {
		return value
	}
	if strings.ContainsAny(value, " /\\") {
		return ""
	}
	return value
}

func historicalDevice(userAgent string) string {
	lower := strings.ToLower(userAgent)
	if strings.Contains(lower, "ipad") || strings.Contains(lower, "tablet") || strings.Contains(lower, "kindle") || strings.Contains(lower, "silk/") || strings.Contains(lower, "android") {
		return "tablet"
	}
	if strings.Contains(lower, "mobile") || strings.Contains(lower, "iphone") || strings.Contains(lower, "ipod") || strings.Contains(lower, "windows phone") {
		return "mobile"
	}
	if userAgent == "" {
		return "unknown"
	}
	return "desktop"
}

func historicalBrowser(userAgent string) string {
	lower := strings.ToLower(userAgent)
	switch {
	case strings.Contains(lower, "edg/"):
		return "edge"
	case strings.Contains(lower, "firefox/"):
		return "firefox"
	case strings.Contains(lower, "chrome/") || strings.Contains(lower, "crios/"):
		return "chrome"
	case strings.Contains(lower, "safari/"):
		return "safari"
	default:
		return "other"
	}
}

func writeHistoricalPV(ctx context.Context, db *sqlstore.DB, values map[historicalPVKey]int) error {
	tx, err := db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer tx.Rollback()
	var query string
	if db.Driver == "mysql" {
		query = `INSERT INTO analytics_historical_pv (day_key,page_path,page_title,source_category,referrer_host,device_type,browser_name,pv_count,imported_at) VALUES (?,?,?,?,?,?,?,?,CURRENT_TIMESTAMP) ON DUPLICATE KEY UPDATE page_title=VALUES(page_title),pv_count=VALUES(pv_count),imported_at=VALUES(imported_at)`
	} else {
		query = `INSERT INTO analytics_historical_pv (day_key,page_path,page_title,source_category,referrer_host,device_type,browser_name,pv_count,imported_at) VALUES (?,?,?,?,?,?,?,?,CURRENT_TIMESTAMP) ON CONFLICT(day_key,page_path,source_category,referrer_host,device_type,browser_name) DO UPDATE SET page_title=excluded.page_title,pv_count=excluded.pv_count,imported_at=excluded.imported_at`
	}
	statement, err := tx.PrepareContext(ctx, query)
	if err != nil {
		return err
	}
	keys := make([]historicalPVKey, 0, len(values))
	for key := range values {
		keys = append(keys, key)
	}
	sort.Slice(keys, func(i, j int) bool {
		if keys[i].day != keys[j].day {
			return keys[i].day < keys[j].day
		}
		return keys[i].page < keys[j].page
	})
	for _, key := range keys {
		if _, err := statement.ExecContext(ctx, key.day, key.page, key.title, key.source, key.host, key.device, key.browser, values[key]); err != nil {
			_ = statement.Close()
			return err
		}
	}
	if err := statement.Close(); err != nil {
		return err
	}
	return tx.Commit()
}

func mustLocationJob(name string) *time.Location {
	location, err := time.LoadLocation(name)
	if err != nil {
		return time.UTC
	}
	return location
}
